### PR TITLE
[Flax] Add Flax inpainting impl

### PR DESCRIPTION
**Pull request by xvjiarui**
Flax version of stable diffusion inpainting model. 

Demo:

```python
import jax
import numpy as np
from flax.jax_utils import replicate
from flax.training.common_utils import shard
import PIL
import requests
from io import BytesIO


from diffusers import FlaxStableDiffusionInpaintPipeline

def download_image(url):
    response = requests.get(url)
    return PIL.Image.open(BytesIO(response.content)).convert("RGB")
img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"

init_image = download_image(img_url).resize((512, 512))
mask_image = download_image(mask_url).resize((512, 512))

pipeline, params = FlaxStableDiffusionInpaintPipeline.from_pretrained(
    # "runwayml/stable-diffusion-inpainting", from_pt=True, dtype=jax.numpy.bfloat16
    "stabilityai/stable-diffusion-2-inpainting", from_pt=True, dtype=jax.numpy.bfloat16
)

prompt = "Face of a yellow cat, high resolution, sitting on a park bench"
prng_seed = jax.random.PRNGKey(0)
num_inference_steps = 50

num_samples = jax.device_count()
prompt = num_samples * [prompt]
init_image = num_samples * [init_image]
mask_image = num_samples * [mask_image]
prompt_ids, processed_masked_images, processed_masks = pipeline.prepare_inputs(prompt, init_image, mask_image)


# shard inputs and rng
params = replicate(params)
prng_seed = jax.random.split(prng_seed, jax.device_count())
prompt_ids = shard(prompt_ids)
processed_masked_images = shard(processed_masked_images)
processed_masks = shard(processed_masks)

images = pipeline(prompt_ids, processed_masks, processed_masked_images, params, prng_seed, num_inference_steps, jit=True).images
images = pipeline.numpy_to_pil(np.asarray(images.reshape((num_samples,) + images.shape[-3:])))
```

### CODE REVIEW

**Comment by xvjiarui** [comment]
cc @patrickvonplaten @patil-suraj @pcuenca

**Comment by patrickvonplaten** [comment]
Very cool! Could we add a test for it?

**Comment by xvjiarui** [comment]
Thx! @patrickvonplaten 

Test added.

File: README.md [hunk 1]
```diff
@@ -284,6 +284,53 @@ output = pipeline(
...
+def download_image(url):
+    response = requests.get(url)
+    return PIL.Image.open(BytesIO(response.content)).convert("RGB")
+img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
+mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"
+
+init_image = download_image(img_url).resize((512, 512))
+mask_image = download_image(mask_url).resize((512, 512))
+
+pipeline, params = FlaxStableDiffusionInpaintPipeline.from_pretrained("xvjiarui/stable-diffusion-2-inpainting")
```
Comment (pcuenca) [inline README.md:310]: Maybe we should also copy the weights to a `flax` branch in `stabilityai/stable-diffusion-2-inpainting`. What do you think @patrickvonplaten @patil-suraj?

**Review by pcuenca — APPROVED** [review]
Great work! I just have a question about copying the weights to the `stabilityai` repo.

**Review by patrickvonplaten — APPROVED** [review]
Very nice work @xvjiarui ! 

@pcuenca @patil-suraj could you also give it a quick review? This looks good to merge for me

File: README.md (hunk 1 above)
Comment (patrickvonplaten) [reply README.md:310]: Given that we will soon deprecate the "branch" behavior, we should probably instead add it to "main", but before doing this we need to be sure that adding flax weights doesn't change the behavior of downloading PyTorch weights. E.g. we should add some tests that if flax weights are present still only PyTorch is downloaded. 

For me ok to merge as is for now.

Comment (pcuenca) [reply README.md:310]: In that case I think it makes sense to merge as is, and then we can migrate the weights when the new system is in place.

File: README.md [hunk 2]
```diff
@@ -284,6 +284,53 @@ output = pipeline(
...
+
+def download_image(url):
+    response = requests.get(url)
+    return PIL.Image.open(BytesIO(response.content)).convert("RGB")
+img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
+mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"
+
+init_image = download_image(img_url).resize((512, 512))
+mask_image = download_image(mask_url).resize((512, 512))
+
```
Comment (patil-suraj) [inline README.md:309]: This is a rather long example, so I'm not sure if we should have this in the main readme, think we could add a doc page for this ? wdyt @pcuenca @patrickvonplaten

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_inpaint.py [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+        if isinstance(mask, Image.Image):
+            mask = [mask]
+
+        processed_images = jnp.concatenate([preprocess_image(img, jnp.float32) for img in image])
+        processed_masks = jnp.concatenate([preprocess_mask(m, jnp.float32) for m in mask])
+        # processed_masks[processed_masks < 0.5] = 0
+        processed_masks = processed_masks.at[processed_masks < 0.5].set(0)
+        # processed_masks[processed_masks >= 0.5] = 1
+        processed_masks = processed_masks.at[processed_masks >= 0.5].set(1)
+
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion_inpaint.py:163]: Nice!

**Review by patil-suraj — APPROVED** [review]
Looks great, thanks a lot for working on this! Just left one nit

File: README.md (hunk 2 above)
Comment (patrickvonplaten) [reply README.md:309]: I think we should refactor the main README quite soon anyways and not have any examples on it. Ok to leave as is for now for me

Comment (patil-suraj) [reply README.md:309]: Okay, in that case all good for me, merging the pr now.

**Pull request merged**